### PR TITLE
 [FLINK-7902] Use TypeSerializer in TwoPhaseCommitSinkFunctions 

### PR DESCRIPTION
**Pull request by aljoscha**

### CODE REVIEW

**Comment by aljoscha** [comment]
R: @kl0u

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 1]
```diff
@@ -362,4 +374,213 @@ public void setContext(Optional<CONTEXT> context) {
...
+				DataInputView source, DataOutputView target) throws IOException {
+			TXN pendingTxn = transactionSerializer.deserialize(source);
+			transactionSerializer.serialize(pendingTxn, target);
+			int numPendingCommitTxns = source.readInt();
+			target.writeInt(numPendingCommitTxns);
+			for (int i = 0; i < numPendingCommitTxns; i++) {
+				TXN pendingCommitTxn = transactionSerializer.deserialize(source);
+				transactionSerializer.serialize(pendingCommitTxn, target);
+			}
+			boolean hasContext = source.readBoolean();
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Here you should also write the `hasContext` to the `target`.

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 2]
```diff
@@ -958,29 +960,34 @@ public int compare(PartitionInfo o1, PartitionInfo o2) {
...
-		public final String transactionalId;
+		final String transactionalId;
 
-		public final long producerId;
+		final long producerId;
 
-		public final short epoch;
+		final short epoch;
 
-		public KafkaTransactionState(String transactionalId, FlinkKafkaProducer<byte[], byte[]> producer) {
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: fixing

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 3]
```diff
@@ -993,14 +1000,162 @@ public String toString() {
 	 * Context associated to this instance of the {@link FlinkKafkaProducer011}. User for keeping track of the
 	 * transactionalIds.
 	 */
-	public static class KafkaTransactionContext {
-		public final Set<String> transactionalIds;
+	static class KafkaTransactionContext {
+		final Set<String> transactionalIds;
 
-		public KafkaTransactionContext(Set<String> transactionalIds) {
+		KafkaTransactionContext(Set<String> transactionalIds) {
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: fixing

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 4]
```diff
@@ -993,14 +1000,162 @@ public String toString() {
...
 			this.transactionalIds = transactionalIds;
 		}
 	}
 
+	static class TransactionStateSerializer extends TypeSerializerSingleton<KafkaTransactionState> {
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: fixing

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 5]
```diff
@@ -993,14 +1000,162 @@ public String toString() {
...
+		public KafkaTransactionState deserialize(
+				KafkaTransactionState reuse,
+				DataInputView source) throws IOException {
+			return deserialize(source);
+		}
+
+		@Override
+		public void copy(
+				DataInputView source, DataOutputView target) throws IOException {
+			boolean hasTransactionalId = source.readBoolean();
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: No, I actually have to do this:
```
			if (hasTransactionalId) {
				target.writeUTF(source.readUTF());
			}
```

But thanks for noticing. :sweat:

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 6]
```diff
@@ -993,14 +1000,162 @@ public String toString() {
...
+			target.writeShort(source.readShort());
+		}
+
+		@Override
+		public boolean canEqual(Object obj) {
+			return obj instanceof TransactionStateSerializer;
+		}
+	}
+
+	static class ContextStateSerializer extends TypeSerializerSingleton<KafkaTransactionContext> {
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: fixing

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 7]
```diff
@@ -362,4 +374,213 @@ public void setContext(Optional<CONTEXT> context) {
 			this.context = context;
 		}
 	}
+
+	/**
+	 * Custom {@link TypeSerializer} for the sink state.
+	 */
+	static final class StateSerializer<TXN, CONTEXT> extends TypeSerializer<State<TXN, CONTEXT>> {
+
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: fixing

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 1 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: fixing

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kl0u) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: Yes, you are right.

**Comment by aljoscha** [comment]
I addressed your comments and also added a new test for the state serialiser.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 8]
```diff
@@ -361,5 +376,247 @@ public void setPendingCommitTransactions(List<TXN> pendingCommitTransactions) {
...
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (o == null || getClass() != o.getClass()) {
+				return false;
+			}
+
+			State<?, ?> state = (State<?, ?>) o;
+
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Not a deal breaker but these if's are a bit too complicated.
Using `Objects.equals()` on each of the components and `&&` would simplify them.

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: I used IntelliJ to generate those. :sweat_smile:

**Comment by aljoscha** [comment]
Thanks for reviewing, @kl0u! 😃

**Pull request closed without merging**